### PR TITLE
quote index columns on oracle, handle all index changes, minor phpdoc…

### PR DESCRIPTION
**Pull request by butonic**
supersedes https://github.com/owncloud/core/pull/27649

It will leave index names unquoted. added a TODO for a future migration from UPPERCASE index identifiers to lowercase identifiers. As index identifiers are suffixed I dont think we will ever hit a reserved identifier ... so I wouldn't bother. Not for esthetic reasons only.

### CODE REVIEW

**Comment by PVince81** [comment]
Code change makes sense, as discussed.

Please modify MigratorTest to enable Oracle tests there for indices.

**Comment by PVince81** [comment]
```
Test\DB\MigratorTest::testDuplicateKeyUpgrade
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'CREATE UNIQUE INDEX i2684056e6ff69bb4d1d3db9f1c190 ON "oc_test_4mdhsfhiim1el" ("id")':

ORA-01452: cannot CREATE UNIQUE INDEX; duplicate keys found

/var/lib/jenkins/workspace/owncloud-core_core_PR-27705-LQ7WTXVIKYSCGVFK3R3KUEAVOLCPWJFSUSND2DREEBTO4YARCGOA@2/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractOracleDriver.php:76
/var/lib/jenkins/workspace/owncloud-core_core_PR-27705-LQ7WTXVIKYSCGVFK3R3KUEAVOLCPWJFSUSND2DREEBTO4YARCGOA@2/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:128
/var/lib/jenkins/workspace/owncloud-core_core_PR-27705-LQ7WTXVIKYSCGVFK3R3KUEAVOLCPWJFSUSND2DREEBTO4YARCGOA@2/lib/composer/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:964
/var/lib/jenkins/workspace/owncloud-core_core_PR-27705-LQ7WTXVIKYSCGVFK3R3KUEAVOLCPWJFSUSND2DREEBTO4YARCGOA@2/lib/private/DB/Migrator.php:199
/var/lib/jenkins/workspace/owncloud-core_core_PR-27705-LQ7WTXVIKYSCGVFK3R3KUEAVOLCPWJFSUSND2DREEBTO4YARCGOA@2/lib/private/DB/Migrator.php:81
/var/lib/jenkins/workspace/owncloud-core_core_PR-27705-LQ7WTXVIKYSCGVFK3R3KUEAVOLCPWJFSUSND2DREEBTO4YARCGOA@2/tests/lib/DB/MigratorTest.php:114
```

Did this test pass locally ?

**Comment by butonic** [comment]
local test:
```
Testing with oci ...
/home/jfd/Repositories/oc/core/lib/composer/phpunit/phpunit/phpunit --configuration phpunit-autotest.xml --coverage-clover autotest-clover-oci.xml --coverage-html coverage-html-oci --log-junit autotest-results-oci.xml ../tests/lib/DB/MigratorTest.php 
PHPUnit 5.7.15 by Sebastian Bergmann and contributors.

Runtime:       PHP 5.6.30-10+deb.sury.org~xenial+2 with Xdebug 2.5.1
Configuration: /home/jfd/Repositories/oc/core/tests/phpunit-autotest.xml

S.......                                                            8 / 8 (100%)

Time: 1.62 minutes, Memory: 54.25MB

There was 1 skipped test:

1) Test\DB\MigratorTest::testDuplicateKeyUpgrade
Does not work yet with Oracle, needs fixing index quoting

/home/jfd/Repositories/oc/core/tests/lib/DB/MigratorTest.php:105

OK, but incomplete, skipped, or risky tests!
Tests: 8, Assertions: 13, Skipped: 1.
```
cool, the PR makes four of five test pass on oracle :)

File: lib/private/DB/OracleMigrator.php [hunk 1]
```diff
@@ -105,17 +116,48 @@ protected function getDiff(Schema $targetSchema, \Doctrine\DBAL\Connection $conn
...
+
+			$tableDiff->addedIndexes = array_map(function(Index $index) {
+				return $this->quoteIndex($index);
+			}, $tableDiff->addedIndexes);
+
+			$tableDiff->changedIndexes = array_map(function(Index $index) {
+				return $this->quoteIndex($index);
+			}, $tableDiff->changedIndexes);
+
+			$tableDiff->removedIndexes = array_map(function(Index $index) {
```
Comment (DeepDiver1975) [inline lib/private/DB/OracleMigrator.php]: isnt removedIndexes a string[] ????

Comment (PVince81) [inline lib/private/DB/OracleMigrator.php]: @butonic please double check and submit a subsequent PR... ideally write a unit test that covers this too

Comment (butonic) [inline lib/private/DB/OracleMigrator.php]: nope, they are Index objects: https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/Comparator.php#L242-L260

**Pull request merged**